### PR TITLE
feat: Move `EdgeVerifierActor` inside `RoutingTableActor`

### PR DESCRIPTION
**Pull request by pmnoxx**
Currently, `PeerManager` is responsible to sending edges to check their signatures `EdgeVerifierActor`.
Then `PeerManager` runs `broadcast_edges_trigger` trigger, which is time consuming.
And then `PeerManager` sends edges to `RoutingTableActor`.

This has a few issues:
- `PeerManager` is spends a bit of time running  `broadcast_edges_trigger`.
- There is an extra structure in `EdgeVerifierActor` which leaks memory and can be removed
- `PeerManager` doesn't care about the edges by itself. This adds extra complexity in the logic.

In the current design:
We create `RoutingTableActor` first then `PeerManagerActor` We pass `Addr<RoutingTableActor>` to `PeerManagerActor`, and we keep references to for both.
- `RoutingTableActor` create `EdgeVerifierActor` and also talks to ` RoutingTableActor`. 
- When new edges are received from network. `PeerManagerActor` sends edges to `EdgeVerifierActors`, reads reply from a `ConcurrentQueue`, and sends them to `RoutingTableActor` for `RoutingTableActor` to update it's storage of `Edges` and updating `RoutingTable`.

That design is a bit compileted:
The final design in 3-4 PRs.
- We create `PeerManagerActor`, `PeerManagerActor` creates `RoutingTableActor`. Then then `RoutingTableActor` created `EdgeVerifierActor`.
- When edges are received from Network. `PeerManagerActor` will send a message with edges to `RoutingTableActor`.
`RoutingTableActor` actor will send edges to `EdgeVerifierActor`, read the response, and update the routing table.

In conclusion:
- `PeerManager` will be less loaded
- We will fix memory leak
- `Code` structure will be easier to understand.

Part 1: `RoutingTableActor` goes inside `RouringTableActor`
- [x] implement ban peer logic
- [x] filter out edges before sending them to `EdgeVerifier`
- [x] Fix tests

Part 2: Offload work from `PeerManagerActor`
- [ ] Move `trigger` from `PeerManager` to `RoutingTableActor`. 

Part 3: Fix memory leak
- [ ] Rewrite how we use `EdgeVerifierActor`

Part 4: Move `RoutingTableActor` fully back to `PeerManagerActor
- [ ] Rewrite how `JsonRpcServer` communicates with `PeerManagerActor`
- [ ] Initially this was done to add support for `JsonRpcServer`, that can be done in a better way.

### CODE REVIEW

**Comment by pmnoxx** [comment]
@matklad I'm not sure where those crashes are coming from. Any ideas?

```


[2021-11-13T09:02:55Z] ---- inactive_recent_edge stdout ----
--
  | [2021-11-13T09:02:55Z] thread 'inactive_recent_edge' panicked at 'System is not running', /var/lib/buildkite-agent/.cargo/registry/src/github.com-1ecc6299db9ec823/actix-rt-2.3.0/src/system.rs:99:21
  | [2021-11-13T09:02:55Z]


```

File: chain/network/src/routing/routing_table_actor.rs [hunk 1]
```diff
@@ -89,6 +101,22 @@ impl RoutingTableActor {
...
+                   at ./tests/cache_edges.rs:44:28
+         8: cache_edges::empty
+                   at ./tests/cache_edges.rs:183:20
+         9: cache_edges::empty::{{closure}}
+                   at ./tests/cache_edges.rs:182:1
+        10: core::ops::function::FnOnce::call_once
+
+
+               */
+        let edge_verifier_pool = SyncArbiter::start(4, || EdgeVerifierActor {});
```
Comment (bowenwang1996) [inline chain/network/src/routing/routing_table_actor.rs:119]: What is this?

Comment (pmnoxx) [reply chain/network/src/routing/routing_table_actor.rs]: @bowenwang1996 That has been fixed. Some tests were not working due to now starting system with `System::new()`

**Comment by pmnoxx** [comment]
@mm-near Can you take a look?

**Comment by mm-near** [comment]
Can you explain a little bit more in the PR description on why it would improve code cleanness ?

**Comment by pmnoxx** [comment]
@mm-near I updated the PR with a better explanation.

File: chain/network/src/routing/routing_table_actor.rs [hunk 2]
```diff
@@ -392,6 +408,11 @@ impl Actor for RoutingTableActor {
     type Context = Context<Self>;
 
     fn started(&mut self, _ctx: &mut Self::Context) {}
+
+    /// Try to gracefully disconnect from active peers.
```
Comment (mm-near) [inline chain/network/src/routing/routing_table_actor.rs:412]: what active peers?

File: chain/network/src/routing/routing_table_actor.rs [hunk 3]
```diff
@@ -408,10 +429,38 @@ impl RoutingTableActor {
 impl Handler<StopMsg> for RoutingTableActor {
     type Result = ();
     fn handle(&mut self, _: StopMsg, _ctx: &mut Self::Context) -> Self::Result {
+        self.edge_verifier_pool.do_send(StopMsg {});
         System::current().stop();
     }
 }
 
+impl Handler<EdgeList> for RoutingTableActor {
```
Comment (mm-near) [inline chain/network/src/routing/routing_table_actor.rs:437]: add comment on what this handler does

File: chain/network/src/peer_manager/peer_manager_actor.rs [hunk 4]
```diff
@@ -1212,33 +1206,22 @@ impl PeerManagerActor {
         );
     }
 
-    fn verify_edges(&mut self, ctx: &mut Context<Self>, peer_id: PeerId, edges: Vec<Edge>) {
+    fn verify_edges(&mut self, _ctx: &mut Context<Self>, peer_id: PeerId, edges: Vec<Edge>) {
```
Comment (mm-near) [inline chain/network/src/peer_manager/peer_manager_actor.rs:1209]: can you mention what this method does?

as it seems to be banning peers that send the invalid edges - but I don't know what happens to the invalid edges themselves. Are we removing them somewhere from the list?

File: chain/network/src/routing/routing_table_actor.rs [hunk 5]
```diff
@@ -408,10 +429,38 @@ impl RoutingTableActor {
...
+        self.edge_verifier_pool
+            .send(msg)
+            .into_actor(self)
+            .map(move |res, act, _| {
+                act.edge_verifier_requests_in_progress -= 1;
+
+                if let Ok(false) = res {
+                    act.peers_to_ban.push(peer_id);
+                }
+                // TODO missing ban peer logic
```
Comment (mm-near) [inline chain/network/src/routing/routing_table_actor.rs:455]: what is the 'ban peer logic' that is missing - can you explain more?

**Review by mm-near — APPROVED** [review]

File: chain/network/src/routing/routing_table_actor.rs (hunk 2 above)
Comment (pmnoxx) [reply chain/network/src/routing/routing_table_actor.rs:412]: "active peers" in this case means peers we are connected to that are in "active_peers" struct.

We should rename it.
https://github.com/near/nearcore/issues/5424

File: chain/network/src/peer_manager/peer_manager_actor.rs (hunk 4 above)
Comment (pmnoxx) [reply chain/network/src/peer_manager/peer_manager_actor.rs:1209]: https://github.com/near/nearcore/pull/5429

File: chain/network/src/routing/routing_table_actor.rs (hunk 5 above)
Comment (pmnoxx) [reply chain/network/src/routing/routing_table_actor.rs:455]: See updated comments in https://github.com/near/nearcore/pull/5429

File: chain/network/src/routing/routing_table_actor.rs (hunk 3 above)
Comment (pmnoxx) [reply chain/network/src/routing/routing_table_actor.rs:437]: See updated comments in https://github.com/near/nearcore/pull/5429

File: chain/network/src/routing/routing_table_actor.rs (hunk 5 above)
Comment (pmnoxx) [reply chain/network/src/routing/routing_table_actor.rs]: Sorry, that's an old comment which was there while I was rewriting the code.
Right now `RoutingTableActor` sends to `PeerManagerActor` periodically, list of peers to ban, for sending us invalid edges.
